sql/hive/src/main/java/org/apache/hadoop/hive/ql/io/orc/VectorizedSparkOrcNewRecordReader.java

```diff
@@ -0,0 +1,317 @@
+  private final org.apache.hadoop.mapred.RecordReader<NullWritable, VectorizedRowBatch> reader;
+  private final int numColumns;
+  private VectorizedRowBatch internalValue;
+  private float progress = 0.0f;
+  private List<Integer> columnIDs;
+
+  private long numRowsOfBatch = 0;
+  private int indexOfRow = 0;
+
+  private final Row row;
+
+  public VectorizedSparkOrcNewRecordReader(
+      Reader file,
+      JobConf conf,
+      FileSplit fileSplit,
+      List<Integer> columnIDs) throws IOException {
+    List<OrcProto.Type> types = file.getTypes();
+    numColumns = (types.size() == 0) ? 0 : types.get(0).getSubtypesCount();
+    this.reader = new SparkVectorizedOrcRecordReader(file, conf,
+      new org.apache.hadoop.mapred.FileSplit(fileSplit));
+
+    this.columnIDs = columnIDs;
+    this.internalValue = this.reader.createValue();
+    this.progress = reader.getProgress();
+    this.row = new Row(this.internalValue.cols, columnIDs);
+  }
+
+  @Override
+  public void close() throws IOException {
+    reader.close();
+  }
+
+  @Override
+  public NullWritable getCurrentKey() throws IOException,
+      InterruptedException {
+    return NullWritable.get();
+  }
+
+  @Override
+  public InternalRow getCurrentValue() throws IOException,
+      InterruptedException {
+    if (indexOfRow >= numRowsOfBatch) {
+      return null;
+    }
+    row.rowId = indexOfRow;
+    indexOfRow++;
+
+    return row;
+  }
+
+  @Override
+  public float getProgress() throws IOException, InterruptedException {
+    return progress;
+  }
+
+  @Override
+  public void initialize(InputSplit split, TaskAttemptContext context)
+      throws IOException, InterruptedException {
+  }
+
+  @Override
+  public boolean nextKeyValue() throws IOException, InterruptedException {
+    if (indexOfRow == numRowsOfBatch && progress < 1.0f) {
+      if (reader.next(NullWritable.get(), internalValue)) {
+        if (internalValue.endOfFile) {
+          progress = 1.0f;
+          numRowsOfBatch = 0;
+          indexOfRow = 0;
+          return false;
+        } else {
+          assert internalValue.numCols == numColumns : "Incorrect number of columns in OrcBatch";
+          numRowsOfBatch = internalValue.count();
+          indexOfRow = 0;
+          progress = reader.getProgress();
+        }
+        return true;
+      } else {
+        return false;
+      }
+    } else {
+      if (indexOfRow < numRowsOfBatch) {
+        return true;
+      } else {
+        return false;
+      }
+    }
+  }
+
+  /**
+   * Adapter class to return an internal row.
+   */
+  public static final class Row extends InternalRow {
+    protected int rowId;
+    private List<Integer> columnIDs;
+    private final ColumnVector[] columns;
+
+    private Row(ColumnVector[] columns, List<Integer> columnIDs) {
+      this.columns = columns;
+      this.columnIDs = columnIDs;
+    }
+
+    @Override
+    public int numFields() { return columnIDs.size(); }
```

Comment:
I've done some profiling of this code (repeated jstack traces on a job that scans the table), and was surprised to see this call taking around 30% of CPU time. Each time .size() is called, the scala sequence to java list conversion has to take place. Other calls that involve columnIDs also took about 9% of CPU time.

I made a change to avoid this conversion, and saw the CPU time spent in these calls reduced to effectively zero. The change was to make a (java) copy of the columnIDs in the constructor of VectorizedSparkOrcNewRecordReader. I replaced `this.columnIDs = columnIDs` at L75 with `this.columnIDs = new ArrayList<>(columnIDs)` and `columnIDs` at L78 with `this.columnIDs`.